pkg/input/input.go

```diff
@@ -0,0 +1,53 @@
+// Package input contains the interfaces for audit logging input
+package input
+
+import (
+	"context"
+	"sync"
+
+	"github.com/go-logr/logr"
+
+	"sandbox.jakexks.dev/cert-manager-audit/pkg/process"
+)
+
+// All registered input adapters are kept here
+var (
+	inputs = make(map[string]Adapter)
+	lock   sync.RWMutex
+)
+
+// An Adapter is an Input factory
+type Adapter interface {
+	New() Input
+	Name() string
+}
+
+// The Input interface should be implemented by all adapters
+type Input interface {
+	// Setup should read any required config, check for dependencies, etc
+	Setup(logr.Logger, process.Func, Config) error
+	// Start should start the adapter listening for incoming events.
+	// It must not block.
+	Start(context.Context) error
+	// Stop should stop accepting any incoming events and store as much as possible.
+	Stop(context.Context) error
+}
+
+// Config is serialized configuration.
+type Config []byte
+
+func Register(adapter Adapter) {
+	lock.Lock()
+	defer lock.Unlock()
+	inputs[adapter.Name()] = adapter
+}
```

Comment:
suggestion: I've often struggled with the sort of k8s-style `Register` workflow like this, with globals - it's used in [cert-manager](https://github.com/jetstack/cert-manager/blob/d24702fb5f85eb37b5dd833e7f3fa48811665d67/pkg/issuer/selfsigned/selfsigned.go#L46-L48) too.

It's not an uncommon pattern and for sure I don't think it _needs_ to change or anything, but I tend to think I'd prefer having an `AdapterManager` (or something named better), where the `AdapterManager` holds the `inputs` map. In `main()` we'd instantiate the manager and explicitly register each type of input.

To me that feels a little easier to test and reason about, if only because no `init()` functions are required for it, avoiding the need for things like:

```go
func init() {
	input.Register(auditAdapter{})
}
```

and instead have:

```go
func main() {
    a := input.AdapterManager{}
	a.Register(input.AuditAdapter{})
}
```

It would probably require either making `auditAdapter` publicly exported or exposing `func NewAuditAdapter() auditAdapter`.

I might be missing something about why the "init workflow" is being used. In any case, it's no big deal but it'd be a pain to change the approach down the road. I'd be interested to hear what you think!